specs/src/shrex/discovery.md

```diff
@@ -0,0 +1,210 @@
+            DHT-->>Disc: Found peer list
+            
+            loop For each found peer
+                Disc->>Host: Connect to peer
+                Host-->>Disc: Connection result
+                
+                alt Connection successful
+                    Disc->>PeerSet: AddPeer(peerID)
+                    Note over Disc: Trigger OnPeersUpdate callback
+                else Connection failed
+                    Note over Disc: Skip peer, continue
+                end
+            end
+            
+            alt Peer set still empty
+                Note over Disc: Wait retry timeout
+            else Peers found
+                Disc-->>App: Return available peers
+            end
+        end
+    end
+```
+
+**Note**: Full nodes (FN) are deprecated in the current protocol implementation but remain part of this specification for completeness as they are still present in the codebase.
+
+The discovery mechanism enables:
+
+- **Light nodes** discover Full and Bridge nodes
+- **Full and Bridge nodes** MAY discover each other for data synchronization
+
+## Protocol Specification
+
+### Discovery Tags
+
+The discovery service uses specific tags to categorize peers by their data availability capabilities:
+
+#### Tag Definitions
+
+**Full Node Tag**: `"full"`
+
+- **Purpose**: Identifies nodes that store recent block data
+- **Capabilities**: Can serve shares for recent heights within the sampling window
+- **Advertisement**: Light nodes MUST NOT advertise under this tag
+- **Discovery**: All node types SHOULD discover peers under this tag
+
+**Archival Node Tag**: `"archival"`
+
+- **Purpose**: Identifies nodes that store historical block data beyond the sampling window
+- **Capabilities**: Can serve shares for all historical heights
+- **Advertisement**: Both Bridge and FULL nodes MAY advertise under this tag in case pruner service is disabled
+- **Discovery**: All node types SHOULD discover peers under this tag
+
+### Discovery Parameters
+
+The discovery service operates with the following configurable parameters:
+
+#### PeersLimit
+
+- **Type**: Integer
+- **Default**: 5
+- **Purpose**: Maximum number of peers to maintain in the limited peer set
+- **Rationale**: Limits resource consumption while ensuring sufficient peer diversity for data availability
+
+#### AdvertiseInterval
+
+- **Type**: Duration
+- **Default**: 1 hour
+- **Purpose**: Interval between peer advertisements to the DHT
+- **Rationale**: Balances network overhead with peer visibility, following DHT best practices
+
+### Discovery Operations
+
+#### Advertisement Process
+
+1. Nodes MUST determine their appropriate tags based on capabilities
+2. Nodes SHOULD periodically advertise their presence under relevant tags
+3. Advertisement includes node's peer ID and network addresses
+
+#### Peer Discovery Process
+
+1. Nodes MUST query the DHT for peers under specific tags
+2. Discovery service maintains a limited set of discovered peers
```

Comment:
one thing is we need to mention connectedness status and that the peers in the limited set must *active connections* - otherwise they must drop the peer for which connectedness changes to anything but connected and discover a new active connection.